### DISCUSSION TITLE
Resolved: Testing

### DISCUSSION THREAD

**Discussion by tcharding** [rust-bitcoin/rust-bitcoin#1683] [Q&A] [answered]
Hi crew,

I have a desire to increase the amount of testing org wide however on multiple occasions when attempting to do so I've either written loads of small tests and then questioned the value (specifically when mutation testing) or else got stuck trying to find things that need testing (specifically with kani).

Two questions please:
1. Tips and/or references so I can become a better tester in general
2. Ideas on things you'd like to see better tested and how (fuzz, kani, mutation)

Thanks

**Comment by dpc** [comment, marked as answer]
One approach to improving tests is:

Imagine you're being paid for infiltrating project and need to find a way to introduce sneaky change with a important but easy to overlook bug. Think of what could that be. Embrace the evil inside you. Let your shadow take control.

Once you think of something, introduce it locally and run tests. If they catch the bug - tests get a point. If they don't - you get a point. Make a commit that passes yet introduces a bug, then another commit with a test that catches it and a fix. 

Repeat until you run out of ideas or time.

**Reply by tcharding** [reply]
Cool, cheers.

**Comment by apoelstra** [comment]
I think probably mutation testing is the lowest hanging fruit here. I'd like to get all of the libs, org-wide, to pass mutagen.

**Comment by tcharding** [comment]
Interestingly it struck me this morning after sleeping on dpc's comment that what he suggests _is_ mutation testing, just done manually. I got psyched by this realisation because it gives me a good path forward, use mutagen on all non-trivial functions then go back over and attack each function as described by @dpc (since I now know exactly how mutagen mutates and what it does not do).

**Reply by dpc** [reply]
I think it's fair to say it's "human guided mutation testing".

I actually only know mutation testing in theory, and never actually implemented it using some framework or tooling. I'd be interested in looking how it works in practice for rust-bitcoin.

**Reply by tcharding** [reply]
I've been messing with `mutagen` and while I don't remember the exact list of mutations after a while one gets a pretty good idea of what is going to break when a function is tested. I've found testing with it really quite enjoyable because its forced be to understand the logic in functions much more deeply that just reading them or throwing up a few quick tests. Its a mad time sink though.
